allennlp/training/metrics/bleu.py

```diff
@@ -0,0 +1,138 @@
+from collections import Counter
+import math
+from typing import Iterable, Tuple, Dict, Set
+
+from overrides import overrides
+import torch
+
+from allennlp.training.metrics.metric import Metric
+
+
+@Metric.register("bleu")
+class BLEU(Metric):
+    """
+    Bilingual Evaluation Understudy (BLEU).
+
+    BLEU is a common metric used for evaluating the quality of machine translations
+    against a set of reference translations. See Papineni et. al.,
+    "BLEU: a method for automatic evaluation of machine translation", 2002.
+
+    Parameters
+    ----------
+    ngram_weights : ``Iterable[float]``, optional (default = (0.25, 0.25, 0.25, 0.25))
+        Weights to assign to scores for each ngram size.
+
+    Notes
+    -----
+    This implementation only considers a reference set of size 1, i.e. a single
+    gold target sequence for each predicted sequence.
+    """
+
+    def __init__(self,
+                 ngram_weights: Iterable[float] = (0.25, 0.25, 0.25, 0.25)) -> None:
+        self._ngram_weights = ngram_weights
+        self._precision_matches: Dict[int, int] = Counter()
+        self._precision_totals: Dict[int, int] = Counter()
+        self._prediction_lengths = 0
+        self._reference_lengths = 0
+
+    @overrides
+    def reset(self) -> None:
+        self._precision_matches = Counter()
+        self._precision_totals = Counter()
+        self._prediction_lengths = 0
+        self._reference_lengths = 0
+
+    @staticmethod
+    def _ngrams(tensor: torch.LongTensor,
+                ngram_size: int,
+                exclude_indices: Set[int] = None) -> Dict[Tuple[int, ...], int]:
+        ngram_counts: Dict[Tuple[int, ...], int] = Counter()
+        if ngram_size > tensor.size(-1):
+            return ngram_counts
+        for start_position in range(ngram_size):
+            for tensor_slice in tensor[start_position:].split(ngram_size, dim=-1):
+                if tensor_slice.size(-1) < ngram_size:
+                    break
+                ngram = tuple(x.item() for x in tensor_slice)
+                if exclude_indices and any(x in ngram for x in exclude_indices):
```

Comment:
This shouldn't really matter given the size of the ngrams and `exclude_indices`, but could you swap `exclude_indices` and `ngram` so that `exclude_indices` has a lookup performed against it instead of being iterated over? Slightly more intuitive that way.